parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java

```diff
@@ -0,0 +1,316 @@
+    public <T> Boolean notStartsWith(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on startsWith
+      return ROWS_MIGHT_MATCH;
+    }
+
+    private BloomFilter bloomById(int id) {
+      if (bloomCache.containsKey(id)) {
+        return bloomCache.get(id);
+      } else {
+        ColumnChunkMetaData columnChunkMetaData = columnMetaMap.get(id);
+        BloomFilter bloomFilter = bloomReader.readBloomFilter(columnChunkMetaData);
+        if (bloomFilter == null) {
+          throw new IllegalStateException("Failed to read required bloom filter for id: " + id);
+        } else {
+          bloomCache.put(id, bloomFilter);
+        }
+
+        return bloomFilter;
+      }
+    }
+
+    private <T> boolean shouldRead(PrimitiveType primitiveType, T value, BloomFilter bloom, Type type) {
+      long hashValue = 0;
+      switch (type.typeId()) {
+        case INTEGER:
+        case DATE:
+          hashValue = bloom.hash(((Number) value).intValue());
+          return bloom.findHash(hashValue);
+        case LONG:
+        case TIME:
+        case TIMESTAMP:
+          hashValue = bloom.hash(((Number) value).longValue());
+          return bloom.findHash(hashValue);
+        case FLOAT:
+          hashValue = bloom.hash(((Number) value).floatValue());
+          return bloom.findHash(hashValue);
+        case DOUBLE:
+          hashValue = bloom.hash(((Number) value).doubleValue());
+          return bloom.findHash(hashValue);
+        case STRING:
+          hashValue = bloom.hash(Binary.fromCharSequence((CharSequence) value));
+          return bloom.findHash(hashValue);
+        case BINARY:
+        case FIXED:
+          hashValue = bloom.hash(Binary.fromConstantByteBuffer((ByteBuffer) value));
+          return bloom.findHash(hashValue);
+        case DECIMAL:
+          BigDecimal decimalValue = (BigDecimal) value;
+          switch (primitiveType.getPrimitiveTypeName()) {
+            case INT32:
+              hashValue = bloom.hash(decimalValue.unscaledValue().intValue());
+              return bloom.findHash(hashValue);
+            case INT64:
+              hashValue = bloom.hash(decimalValue.unscaledValue().longValue());
+              return bloom.findHash(hashValue);
+            case FIXED_LEN_BYTE_ARRAY:
+            case BINARY:
+              DecimalMetadata metadata = primitiveType.getDecimalMetadata();
+              int scale = metadata.getScale();
+              int precision = metadata.getPrecision();
+              byte[] requiredBytes = new byte[TypeUtil.decimalRequiredBytes(precision)];
+              byte[] binary = DecimalUtil.toReusedFixLengthBytes(precision, scale, (BigDecimal) value, requiredBytes);
+              hashValue = bloom.hash(Binary.fromConstantByteArray(binary));
+              return bloom.findHash(hashValue);
+          }
+        case UUID:
+          hashValue = bloom.hash(Binary.fromReusedByteArray(UUIDUtil.convert((UUID) value)));
```

Comment:
I don't think this byte array is reused.

Reply:
Changed to `fromConstantByteArray`. Thanks!